src/value_generators/src/DocumentGenerator.cpp

```diff
@@ -745,6 +747,73 @@ class StringGenerator : public Generator<std::string> {
     const size_t _alphabetLength;
 };
 
+/** `{^ChooseFromDataset:{...}` */
+class RandomDataFromDataset : public Generator<std::string> {
+
+public:
+    RandomDataFromDataset(const Node& node, GeneratorArgs generatorArgs)
+        : _rng{generatorArgs.rng},
+          _id{generatorArgs.actorId},
+          _path{node["path"].maybe<std::string>().value()},
+          _pathLength{_path.size()}
+      {
+
+
+        if (_pathLength <= 0) {
+            BOOST_THROW_EXCEPTION(InvalidValueGeneratorSyntax(
+            "RandomDataFromDataset requieres non-empty path"));
+            }
+
+            /* To avoid reading and storing the datasets multiple times
+               we check if the path already exists in the map (as key). If
+               it doesn't exist, we open the file and store its contents. 
+               Section locked with a mutex */
+            
+            _dataset_mutex.lock();
+            if (_dataset.count(_path) == 0) {
+                std::cout << "I enter to open " << _path << std::endl;
+                _ifs.open(_path, std::ifstream::in);
+                if (_ifs.is_open()){
+                    while(std::getline(_ifs, _current_line))
+                    {
+                        // Ignore emtpy lines
+                        if (_current_line.size()!=0){
+                            _dataset[_path].push_back(_current_line);
+                        }
+                    }
+                }
+            
+            else {
+                _dataset_mutex.unlock();
+                BOOST_THROW_EXCEPTION(InvalidValueGeneratorSyntax(
+                    "The specified file for RandomDataFromDataset cannot be opened or it does not exist")); }
+
+            if (_dataset[_path].size() == 0) {
+                _dataset_mutex.unlock();
+                BOOST_THROW_EXCEPTION(InvalidValueGeneratorSyntax(
+                    "The specified file for RandomDataFromDataset is empty")); }
+                }
+
+            _dataset_mutex.unlock();
+        }
+
+    std::string evaluate() {   
+        auto distribution = boost::random::uniform_int_distribution<size_t>{0, _dataset[_path].size() - 1};
+        return _dataset[_path][distribution(_rng)];
+    }
+    
+private:
+    DefaultRandom& _rng;
+    ActorId _id;
+    std::string _path;
+    const size_t _pathLength;
+    std::ifstream _ifs;
+    std::vector<std::string> _dataSetContent;
+    std::string _current_line;
+    static inline std::unordered_map<std::string, std::vector<std::string>> _dataset;
```

Comment:
Are you using the full path as the key here? That probably makes sense. My first reaction though was to put in an explicit key. That said, I can't come up with a valid reason why there would ever be a non 1-1 relationship from keys to paths.

Reply:
Yes, the key is the path, since there won't be two different files with the same path. I think this method is simple enough and it works for what we need, since we already have the path and it is unique.